File: client/src/types/notification.ts
// Notification DTOs. The server stores the type and its two values; the text is rendered here so
// it follows the reader's language.

import type { PageParams } from "@/types/common";

export const NotificationType = [
  "APPLICATION_ACCEPTED",
  "APPLICATION_REJECTED",
  "INTERVIEW_SCHEDULED",
] as const;
export type NotificationType = (typeof NotificationType)[number];

export interface NotificationDTO {
  id: number;
  type: NotificationType;
  subject: string | null; // the job title
  context: string | null; // the company
  link: string | null;
  createdAt: string;
  readAt: string | null; // null = still unread, so still on the dashboard
}

export type NotificationFilters = PageParams;

File: server/src/main/java/dev/sabti/alumni_connect/notification/Notification.java
package dev.sabti.alumni_connect.notification;

import dev.sabti.alumni_connect.auth.entities.User;
import dev.sabti.alumni_connect.shared.BaseEntity;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

// One thing that happened to one user. The sentence is rendered client-side from the type, so a
// notification reads in whichever language the reader has picked.
@Entity
@Table(name = "notifications")
@Getter @Setter
@AllArgsConstructor @NoArgsConstructor
@EqualsAndHashCode(callSuper = true)
public class Notification extends BaseEntity {
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private NotificationType type;

    // What it's about (a job title) and where (a company).
    private String subject;

    private String context;

    // Relative path the client navigates to when it's clicked.
    private String link;

    // Null until clicked: unread is what the dashboard shows.
    private LocalDateTime readAt;
}
